src/mixings/conditional_mixing.h

```diff
@@ -0,0 +1,32 @@
+#ifndef BAYESMIX_MIXINGS_CONDITIONAL_MIXING_H_
+#define BAYESMIX_MIXINGS_CONDITIONAL_MIXING_H_
+
+#include "base_mixing.h"
+
+class ConditionalMixing : public BaseMixing {
+ public:
+  bool is_conditional() const override { return true; }
+
+  //! Mass probability for choosing an already existing cluster
+  double mass_existing_cluster(const unsigned int n, const bool log,
+                               const bool propto,
+                               std::shared_ptr<AbstractHierarchy> hier,
+                               const Eigen::RowVectorXd &covariate =
+                                   Eigen::RowVectorXd(0)) const override {
+    throw std::bad_function_call(

```

Comment:
I guess that having to throw errors in derived classes is a poor design choice.

The cleanest design I can think of is to have:
```
class BaseMixing {
 public:
   virtual ~BaseMixing() {}
};

class MarginalMixing : virtual public BaseMixing {
 public:
   virtual double mass_existing_cluster(...);
   virtual double mass_new_cluster(...);
};

class ConditionalMixing: virtual public BaseMixing {
 public:
   virtual Eigen::VectorXd get_weights(...);
};
```

Then, in `MarginalAlgorithm` we store an `std::shared_ptr<MarginalMixing>` while in `ConditionalAlgorithm` we store a `ConditionalMixing`.
To keep a uniform API across all algorithms, we can define `BaseAlgorithm::set_mixing(const std::shared_ptr<BaseMixing> mixing_)` as virutal, and then `MarginalAlgorithm` we can write
```
MarginalAlgorithm::set_mixing(std::shared_ptr<BaseMixing> mixing_) {
  mixing = std::dynamic_pointer_cast<MarginalMicing>(mixing_);
}
```
same goes for ConditionalAlgorithm.

In this way, we can also have the "diamond of death", so that some mixings can be both marginal and conditional (for instance the finite-dimensional dirichlet, or the finite dimensional dirichlet with a random number of atoms, is both)

Reply:
Sounds great!

Reply:
Upon second thoughts, we actually do a lot of stuff with `mixing` in the base algorithm, such as checking it is non-null, using it in `get_state_as_proto()`, etc. More in general, it might not be a good idea not to have it in the base class if it is part of the API, albeit in different forms depending on the type of algorithm. We could create some type of internal getter/wrapper in both intermediate classes (marginal and conditional) that performs a dynamic cast to the appropriate type, similarly to what we were doing with Hierarchies before the refactoring. Is there any easier solution?

Reply:
That's a good point!
We can actually have a CRTP also for the Algorithms if that makes life simpler, or just go with having some 'casting' methods in conditional and marginal algorithms. Final possibility is to have both a pointer to BaseMixing and a pointer to MarginalMixing, that point to the same object and use whichever pointer we need

Reply:
> Final possibility is to have both a pointer to BaseMixing and a pointer to MarginalMixing, that point to the same object and use whichever pointer we need

This is an excellent idea! We can try to apply CRTP later if we think it would improve things. For now I'd stick to this one.

Reply:
Great!